[Cmake] implement (some) ninja support

fuzzard opened the pull request:
## Description
Implements the ability to use ninja to build (some) target projects

## Motivation and context
With some recent changes to adaptive build types, @enen92 alerted me to the fact ninja no longer worked. It only actually worked out of sheer luck really, but this should bring it back for (some) platforms.

Cmake internal lib builds have been standardised to use an IMPORTED library target (eg CrossGUID::CrossGUID). This is done as externalproject_add targets cant easily be added to an imported target. We worked around this previously by using a property list to force add the  externalproject target to libkodi to build. This is a more "Cmake" way i believe

## How has this been tested?
All the following was carried out with the following internal build flags
```
-DENABLE_INTERNAL_FSTRCMP=OFF -DENABLE_INTERNAL_TAGLIB=ON -DENABLE_INTERNAL_SPDLOG=ON -DENABLE_INTERNAL_FMT=ON -DENABLE_INTERNAL_CROSSGUID=ON
```
Build xcode ios project
Build macos x86_64 Debug/Release Unix Makefiles
Build macos x86_64 Debug/Release Ninja
Build aarch64 android Debug Ninja (libkodi.so target, apk target doesnt work)

## What is the effect on users?
N/A

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Commit:
[cmake] Crossguid use CrossGUID::CrossGUID target for depends

Implement CrossGUID::CrossGUID target to allow adding dependency target for
ENABLE_INTERNAL_CROSSGUID.

Commit:
[cmake] Fmt use Fmt::Fmt target for depends

Implement Fmt::Fmt target to allow adding dependency target for
ENABLE_INTERNAL_FMT.

Commit:
[cmake] introduce Spdlog::Spdlog target

Allow ENABLE_INTERNAL_SPDLOG to set a build target spdlog and add it as a dependency
to the imported target. The imported target is then just a blanket dep include to cover
either system or internal lib

Commit:
[cmake] tidy up TagLib::TagLib target

Commit:
[cmake] FFMPEG set FFMPEG_LIB_TYPE for ENABLE_INTERNAL_FFMPEG build

Commit:
[cmake] BUILD_DEP_TARGET set BYPRODUCT for cmake based builds

Required to make ninja play nice when using ENABLE_INTERNAL_<LIB>

Commit:
[tools/depends] cmakebuildsys allow ninja Generator to used/created

use the following to create a Ninja based build project
  make -C tools/depends/target/cmakebuildsys GEN=Ninja

This implements the single config ninja generator (ie fixed BUILD_TYPE) in cmake

Commit:
[cmake] minor order change to find module debug postfix set

Commit:
[cmake] Remove default DEBUG_POSTFIX for ArchSetup

remove arch based DEBUG_POSTFIX. turns out we are better to have the individual
lib builds to define a postfix when appropriate.

Commit:
[cmake] Windows increase minimum required cmake version 3.20

Generator expressions for Multi-config generators (ie VS projects) require
cmake 3.20.0 as a minimum. VS 2017 comes with version 3.20.21032501-MSVC_2
bundled.

fuzzard commented:
Ok, have reworked the byproduct stuff. relocated the debug postfix handling to the BUILD_DEP_TARGET. Its a little more relevant to that stuff than the VERSION file parsing.

One thing ive found, generator expressions for VS dont work with cmake < 3.20. Ive increased the cmake version requirement for windows from 3.15 to 3.20. Of note, VS 2019 (community edition) ships with cmake 3.20, so theres not really a need to use a standalone cmake install.

The generator expressions are only used for multi-config based projects (ie Xcode, VS). So this doesnt affect linux users at all, so cmake version requirement can stay where it is. i'll mention ninja does have a multiconfig variant, short answer is i havent tested it, assume it doesnt work, dont be stupid and try to use it.

I need to get approval to remove (or update PATH) to get rid of the standalone cmake 3.15.5 install we have on our jenkins agents. Until i get that confirmation/action, the windows agents will all fail.

Commit:
[cmake] FindFmt module enable_internal to check existing lib

enables the ENABLE_INTERNAL_FMT path to test for existing lib/package and if it meets
the version supplied by FMT-VERSION, we dont need to rebuild.

Commit:
[cmake] FindSpdlog ENABLE_INTERNAL to test existence

Allows ENABLE_INTERNAL_SPDLOG path to test for existence of package/lib meeting
SPDLOG-VERSION version requirements, and not build if found.
Also force ENABLE_INTERNAL_FMT on if we are doing ENABLE_INTERNAL_SPDLOG due to
fmt version requirements for spdlog

fuzzard commented:
@wsnipex sorry to give you grief. Ive gone against my one rule (keep the PR's simple and small). 
Ive reworked the FindFmt and FindSpdlog in 2 new commits. Can you have a look again, particularly at the last 2 commits.

Essentially on some systems (@enen92 ) its still picking up system includes for fmt when using enable_internal_spdlog and fmt. 
The changes now force internal fmt on when internal spdlog is done. Its just easier due to fmt backward incompatibilities.
Both internal spdlog and fmt will now do a find_package call looking for an existing install, and test against our version supplied in the VERSION file in tools/depends. if its older, it'll build internally, if not, it will just populate the include/lib variables and not create the build target.

This would eventually have to have been done, but what this allows is a system to build internal libs once (gets installed into the DEPENDS_PATH location), and then on subsequent cmake generation/run, it'll find those prebuilt libs and just skip building, as long as the version is up to date.
With cmake built libs (ie fmt/spdlog) we can use the cmake config files created/installed more easily on all platforms (ie windows) for the find_package without having to deal with pkgconfig. When we get to autoconf based libs this will get more difficult for windows, but baby steps.
Im not sure if we want to do a force rebuild, but really this should be the most common path. build once, reuse unless VERSION file increases. A force rebuild adds extra complexity, when a user can just clean out the DEPENDS_PATH location to get a fresh start. 

This also is essentially how i see the system handling rechi's comment at https://github.com/xbmc/xbmc/pull/21311#issuecomment-1109457660

wsnipex commented:
to make reusing stuff from DEPENDS_PATH work for non-depends builds, we need to add $DEPENDS_PATH to CMAKE_PREFIX_PATH

fuzzard commented:
Yeah, I was looking at that last night. Trying to work out how/where it if https://github.com/xbmc/xbmc/blob/139e39995e9ffe38b583aef84dde0b1f03fa3deb/CMakeLists.txt#L15 is ever used. As far as I can tell, it's not. So I'd look at changing that to depends_path essentially. Would probably have to relocate it a little later (probably after pathsetup)

wsnipex commented:
that's something for addons afaik.
Maybe something like https://github.com/wsnipex/xbmc/blob/57e4a903a7a39321ed14ac691e0fe3b83f4b71bc/cmake/scripts/common/ArchSetup.cmake#L56-L58 is enough. It'd be nicer to only do that if at least 1 ENABLE_INTERNAL_XX is enabled, but I have no idea how to best detect that

enen92 commented on CMakeLists.txt:
```diff
@@ -1,7 +1,7 @@
-  # Version 3.15 is required to use "PREPEND" for dependencies
-  cmake_minimum_required(VERSION 3.15)
+  # Version 3.20 is required for multi-config generator expressions to work
+  cmake_minimum_required(VERSION 3.20)
```
it will need changes to the build guides and jenkins workers it seems

fuzzard replied:
Doc change is done. It's windows only.

Commit:
[cmake] pass cmake_prefix_path through to internaldep build function

Prepend depends_path to cmake_prefix_path for find_package/program/etc calls to
search our depends_path first
Pass through cmake_prefix_path to internal build functions for cmake based build systems

Commit:
[cmake] refactor BUILD_DEP_TARGET to utilise generator expressions

wsnipex commented:
spdlog/fmt seems ok now, but internal ffmpeg is broken. FFmpeg is built, but the final link fails. On 2nd cmake run ffmpeg will be picked up from DEPENDS_PATH as "external" and it builds fine.

fuzzard commented:
Is this ffmpeg failure when using ninja, or as a makefile based build? Or both?

Commit:
[cmake] ffmpeg-link-wrapper update to match DEPENDS_PATH usage for lib build location

fuzzard commented:
OK, the ffmpeg issue should be fixed now. Missed an update to the wrapper script to use DEPENDS_PATH for PKG_CONFIG_PATH.

Hopefully thats the last for this PR. Still trying to get the windows jenkins slaves "fixed", but thats not a PR issue

howie-f reviewed (approved):
tested on Debian 11 with cmake 3.18.4
configurations Release and Debug work with both generators, Unix Makefiles and Ninja.
apart from the windows build slaves i consider this ready now.
cheers @fuzzard. i know you now curse day and hour you first touched cmake's files :)

(empty commit https://github.com/xbmc/xbmc/pull/21463/commits/6ff10075a0115268414cadbace3e6ef7ed4872d4 can be dropped)

wsnipex commented on cmake/modules/FindFFMPEG.cmake:
```diff
@@ -109,12 +109,12 @@ macro(buildFFMPEG)
-  avutil=`PKG_CONFIG_PATH=${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/lib/pkgconfig ${PKG_CONFIG_EXECUTABLE} --libs --static libavutil`
-  swscale=`PKG_CONFIG_PATH=${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/lib/pkgconfig ${PKG_CONFIG_EXECUTABLE} --libs --static libswscale`
-  swresample=`PKG_CONFIG_PATH=${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/lib/pkgconfig ${PKG_CONFIG_EXECUTABLE} --libs --static libswresample`
+  avformat=`PKG_CONFIG_PATH=${DEPENDS_PATH}/lib/pkgconfig ${PKG_CONFIG_EXECUTABLE} --libs --static libavcodec`
```
I know it was wrong before, but this should be `avformat`
and `avcodec` below

fuzzard replied:
Are you sure? testing on osx i get this 

```
MacBook-Pro-2 ~ % PKG_CONFIG_PATH=/Users/Shared/xbmc-depends/macosx12.3_x86_64-target-debug/lib/pkgconfig /Users/Shared/xbmc-depends/arm-darwin21.4.0-native/bin/pkg-config --libs --static avformat
Package avformat was not found in the pkg-config search path.
Perhaps you should add the directory containing `avformat.pc'
to the PKG_CONFIG_PATH environment variable
No package 'avformat' found
MacBook-Pro-2 ~ % PKG_CONFIG_PATH=/Users/Shared/xbmc-depends/macosx12.3_x86_64-target-debug/lib/pkgconfig /Users/Shared/xbmc-depends/arm-darwin21.4.0-native/bin/pkg-config --libs --static libavformat
-L/Users/Shared/xbmc-depends/macosx12.3_x86_64-target-debug/lib -lavformat -lm -lbz2 -lz -lgnutls -lgmp -ldl -lhogweed -lgmp -lnettle -lavcodec -lm -llzma -ldav1d -lz -framework AudioToolbox -pthread -framework VideoToolbox -framework CoreFoundation -framework CoreMedia -framework CoreVideo -framework CoreServices -lswresample -lm -lavutil -pthread -lm -framework VideoToolbox -framework CoreFoundation -framework CoreMedia -framework CoreVideo -framework CoreServices
brent@Brents-MacBook-Pro-2 ~ % 
```

wsnipex replied:
sorry, ofc it's libavformat.
point was:
 **avformat**`=`PKG_CONFIG_PATH=${DEPENDS_PATH}/lib/pkgconfig ${PKG_CONFIG_EXECUTABLE} --libs --static **libavformat**
vs
avformat=PKG_CONFIG_PATH=${DEPENDS_PATH}/lib/pkgconfig ${PKG_CONFIG_EXECUTABLE} --libs --static _libavcodec_

enen92 reviewed (approved):
Been using this as my base branch for quite some days :) thanks much for the fix

Commit:
[cmake] Findffmpeg fix reversal of var names for link-wrapper script

enen92 commented:
Jenkins build this please

fuzzard commented:
No point in that, working through jenkins things still.

fuzzard commented:
Jenkins build this please
